---
title: "GEGVIC-NGS"
author: "Oriol Arqu√©s"
date: '`r format(Sys.Date(),"%e de %B, %Y")`'
output: 
    html_document:
        toc: yes
params:
    counts: !r system.file("extdata", "input_ge_module.csv", package = "GEGVIC")
    genes_id: 'entrezgene_id'
    metadata: !r system.file("extdata", "metadata_ge_module.csv", package = "GEGVIC")
    colors: !r c('black', 'orange')
    design: 'Response'
    ref_level: !r c('Response', 'Non_Responders')
    shrink: 'none'
    fold_change: 1.5
    p.adj: 0.05
    gmt: !r system.file("extdata", "c7.all.v7.2.symbols.gmt", package = "GEGVIC")
    gsea_pvalue: 1
    biomart: !r system.file("extdata", "ensembl_biomart_GRCh38_p13.csv", package = "GEGVIC")
    muts: !r system.file("extdata", "input_gv_module.csv", package = "GEGVIC")
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r load}
library(GEGVIC)
library(clusterProfiler)

counts <- read.csv(params$counts, header = TRUE, sep = ',')
metadata <- read.csv(params$metadata, header = TRUE, sep = ',')
gmt <- read.gmt(params$gmt)
biomart <- read.csv(params$biomart, header = TRUE, sep = ',')

#gv_module
muts <- read.csv(params$muts, header = TRUE, sep = ',')

```


```{r pca}
# Get data PCA
    ge_pca(counts = counts,
                  genes_id = params$genes_id,
                  metadata = metadata,
                  design = params$design,
                  colors = params$colors)

    
```


```{r de}
# Run differential gene expression analysis
    results.dds <- ge_diff_exp(counts = counts,
                               genes_id = params$genes_id,
                               metadata = metadata,
                               design = params$design,
                               ref_level = params$ref_level,
                               shrink = params$shrink)
```


```{r annot}
# Annotate gene symbols
    annot.res <- ge_annot(results_dds = results.dds,
                          genes_id = params$genes_id,
                          biomart = biomart)

```


```{r volcano}
# Create volcano plot
    ge_volcano(annot_res = annot.res,
               fold_change = params$fold_change,
               p.adj = params$p.adj)
```


```{r gsea}
# Obtain GSEA results
    ge_gsea(annot_res = annot.res,
            gmt = gmt,
            gsea_pvalue = params$gsea_pvalue)

```


## GV_module 
```{r mut_signatures}
gv_mut_signatures(muts = muts,
                   metadata = metadata,
                   response = Response,
                   gbuild = 'BSgenome.Hsapiens.UCSC.hg19',
                   mut_sigs = 'COSMIC_v2_SBS_GRCh37',
                   tri.counts.method = 'default',
                   colors = c('black', 'orange'))
```


